### ISSUE TITLE
Proposal: Connection Resetting Middleware for SET search_path

### ISSUE THREAD

**Issue by Amnesthesia** [rails-on-services/apartment#302]
## Steps to reproduce
Use PgBouncer or an equivalent (e.g AWS RDS Proxy) in front of your database with Rails

## Expected behavior
PgBouncer to be able to reuse connections efficiently

## Actual behavior
Near 1-1 connections because connections become session-pinned when `SET` is encountered

## System configuration

* Database: postgres 16.4 + AWS RDS Proxy

* Apartment version: latest (HEAD)

* Apartment config (in `config/initializers/apartment.rb` or so):

  * `use_schemas`: true

* Rails (or ActiveRecord) version: 7.0.8.6

* Ruby version: 3.3.5

## Proposed solution

```ruby
class Apartment::ResetConnection
  def initialize(app)
    @app = app
  end

  def call(env)
    @app.call(env)
  ensure
    # Reset search path to allow connections to be unpinned if the SET pinned them
    Apartment.connection.execute('RESET search_path')
  end
end
```

Alternatively, this could be done in the postgres adapter:
```ruby
  class PostgresqlAdapter < AbstractAdapter
      ....
      #   Reset schema search path to the default schema_search_path
      #
      #   @return {String} default schema search path
      #
      def reset
        @current = default_tenant
        Apartment.connection.schema_search_path = full_search_path
        Apartment.connection.execute('RESET search_path')
      end

      ....
  end
```

I'm not super familiar with how this works in Apartment, so thought I'd open an issue with the proposal rather than a PR. What do you think, would this be possible? Would it cause any issues?

**Comment by mnovelo** [comment]
hmmm, I'd think that a SET search path would be issued whenever a query was about to be sent through a connection, so it doesn't matter what the previous value of search path was. This is why I don't think there's been much done to issue an additional SQL statement to reset the search path on the connection. Can you tell me more about your use case? When are you specifying a tenant, i.e., the search path?

**Comment by Amnesthesia** [comment]
@mnovelo We set the tenant during authentication, i.e after parsing the JWT token and authenticating a user. So, I agree that it doesn't matter what the previous value of search path was, but when RDS Proxy comes across a `SET` command (SET search_path, SET timezone, doesn't matter). Here's some info on that: https://jpcamara.com/2023/04/12/pgbouncer-is-useful.html

> Amazon has a similar tool to PgBouncer called RDS Proxy, and it has a feature called “[connection pinning](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/rds-proxy-managing.html#rds-proxy-pinning)”. If it detects a statement that is incompatible with transaction mode, it will automatically hold that connection for that client for the duration of their session.

This is both highly useful and simultaneously problematic. It means query behavior is consistent with your expectations (🙌🏼) but also that you can silently kill all concurrency benefits (😑). If enough queries are run that trigger connection pinning, all of a sudden you may throttle your [throughput](https://jpcamara.com/2023/04/12/pgbouncer-is-useful.html#throughput). But it does give you an escape hatch for safely running statements which are not transaction compatible without having to jump through any hoops. 

The problem is that this leads to pretty much all connections being session-pinned because of Apartment setting the search path, and not resetting it

<img width="443" alt="Screenshot 2024-11-19 at 11 00 25" src="https://github.com/user-attachments/assets/5bb6cb73-d8d9-4e07-8f5f-ab2f6887ccd0">

**Comment by mnovelo** [comment]
Oh, I see. Then, something like what you're proposing would be helpful for people using either AWS RDS Proxy or PgBouncer. I'd prefer for this to be something that has to be enabled by configuration. This would execute an unnecessary query for those not using AWS RDS Proxy or PgBouncer, so I'd want this behavior disabled by default. I'm open to names for this configuration, like `enforce_search_path_reset` or similar.

**Comment by Amnesthesia** [comment]
@mnovelo I see what you're saying, on the other hand, wouldn't it make sense that the `reset` method actually resets things that it has set, such as search path? Yeah, it executes an additional query, but not a very heavy one, and it really just undoes what it did by setting the search path in the first place. I think it seems like reasonable cleanup?

I think `enforce_search_path_reset ` sounds like a good name for such a configuration, or `reset_search_path_before_disconnect`

**Comment by mnovelo** [comment]
Let's go with `enforce_search_path_reset` with a default of false. While I see what you're saying about cleanup, it's similar to the `tenant_presence_check`, which is also an arguably good practice to have enabled. What we've found in our production environment is that such an additional safeguard was not worth the overhead of an extra query on each request. At this time, I don't see this being a necessity outside of connection pooling solutions like RDS Proxy and PgBouncer, which is why I'm defaulting it to false. I'm happy to default it to true in the future if more people find it beneficial.

**Comment by Amnesthesia** [comment]
@mnovelo When do you think this could be added?

**Comment by mnovelo** [comment]
I'm looking to include it in v4.0.0.alpha1 which I hope to release by EOY with Rails 7.2 and 8 support

**Comment by mnovelo** [comment]
This wasn't included in v3.2.0, but I hope to add such support in v4.0.0. Your help would be greatly appreciated! https://github.com/rails-on-services/apartment/discussions/312

**Comment by Amnesthesia** [comment]
@mnovelo Awesome! 

As for the proposed solutions above, what's your opinion?

For a middleware, I could imagine that if any fibers or threads spawned during the runtime of the request checked out connections and called `switch` to switch into a tenant, then those wouldn't be reset, only `Apartment.connection` would be reset

As for the other proposed solution, adding it to the `reset` method, this would then call `RESET search_path` even if the request isn't necessarily finished yet, and would conflict with the behaviour of setting `search_path` back to `public` or whatever it was before switching.

Are there perhaps any hooks we can hook into, called when a connection is checked back in?

**Comment by mnovelo** [comment]
My current thinking with this is that we check out connections **with** the tenant specified and that a connection's tenant is immutable. Each tenant would have its own connection pool, and there'd be a global connection pool for the maximum number of connections allowed across all tenants. What do you think of this @Amnesthesia ?

**Comment by Amnesthesia** [comment]
@mnovelo That sounds great

**Comment by mnovelo** [comment]
Substantially addressed by the v4 redesign. The pinning here came from v3 issuing a `SET search_path` on a shared connection on every tenant switch — exactly the transaction-mode-incompatible statement that pins the connection under PgBouncer/RDS Proxy.

v4 is pool-per-tenant with immutable per-pool connection config: the search_path is established once when a tenant's pool connects, not re-`SET` on every request switch. That removes the per-request `SET` churn that drove near-1:1 pinning. The `enforce_search_path_reset` config we discussed here was therefore dropped — v4's model makes it unnecessary.

Honest caveat: a search_path established via `SET` at connection setup can still pin that pooled connection once (just not per request). If you run v4 behind RDS Proxy and still see pathological pinning, please open a fresh issue with proxy pinning metrics — that'd be a distinct, narrower problem than the per-switch churn this issue is about. Closing as superseded by v4.

**Comment by mnovelo** [comment]
Follow-up: the residual gap — fully eliminating the one-time connection-setup `SET` via the libpq `options` path — is now tracked in #438.

**Comment by Amnesthesia** [comment]
How does pool per tenant work? One connection pool per tenant? Does that mean with hundreds of tenants we'd have lots of connections?

**Comment by mnovelo** [comment]
Yes — one pool per tenant, keyed `tenant:role` (so `writing` + `reading` is two). But tenant count isn't what bounds connections: pools are created lazily per process on first switch, connections inside a pool open on demand, and idle pools are reaped whole after `pool_idle_timeout` (300s). A worker that touches 8 tenants holds 8 pools, not 500.

For a hard number rather than a shape:

```ruby
config.tenant_pool_size = 5
config.max_tenant_connections = 250   # => floor(250 / 5) = 50 tenant pools
config.pool_overflow_policy = :raise  # hard ceiling; the default :evict_idle is soft
```

That budget is per process — fleet-wide is `processes × budget` plus each process's default pool, and the `tenant_pools_live` / `backend_connections` gauges report the actual footprint rather than the ceiling ([Pool Settings](https://github.com/rails-on-services/apartment/blob/main/README.md#pool-settings)). One aside, since this issue started with RDS Proxy: on PostgreSQL it pins sessions for reasons mostly Rails' rather than Apartment's, so it won't cut your connection count either way ([docs/connection-poolers.md](https://github.com/rails-on-services/apartment/blob/main/docs/connection-poolers.md)) — `max_tenant_connections` is the more direct lever.

Worth calibrating against, though: we run 500+ tenants ourselves and are still assessing v4 in staging, with production testing next. So the pool-per-tenant model is sound and covered by tests, but not yet proven in production at that scale — it's still an alpha (`4.0.0.alpha11`). If you're in similar territory, I'd be glad to hear how many distinct tenants one of your processes serves in a five-minute window; that's the number that drives sizing, and it's usually far below the total.

**Comment by Amnesthesia** [comment]
@mnovelo If we have 250 tenants all using the application throughout the day, and 16 puma workers, and we normally have a pool of 3 (3 threads per worker), then should we expect our connection count to balloon from 16x3=**48** to 16x3x250=**12000**? 

We dont really control what worker serves what tenant, its being loadbalanced according to what worker has the smallest backed up puma queue, so any worker could be serving any tenant

We have moved away from RDS proxy since it wasn't doing anything for us and was pointless with the session pinning
